java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java

```diff
@@ -93,119 +98,229 @@
+
+        List<Field> children = null;
+        if (arrowType.getTypeID() == ArrowType.List.TYPE_TYPE) {
+          final JdbcFieldInfo arrayFieldInfo = getJdbcFieldInfoForArraySubType(rsmd, i, config);
+          if (arrayFieldInfo == null) {
+            throw new IllegalArgumentException("Configuration does not provide a mapping for array column " + i);
+          }
+          children = new ArrayList<Field>();
+          final ArrowType childType =
+              getArrowTypeForJdbcField(arrayFieldInfo, config.getCalendar());
+          children.add(new Field("child", FieldType.nullable(childType), null));
+        }
+
+        fields.add(new Field(columnName, fieldType, children));
       }
     }
 
     return new Schema(fields, null);
   }
 
+  /**
+   * Creates an {@link org.apache.arrow.vector.types.pojo.ArrowType}
+   * from the {@link JdbcFieldInfo} and {@link java.util.Calendar}.
+   *
+   * <p>This method currently performs following type mapping for JDBC SQL data types to corresponding Arrow data types.
+   *
+   * <ul>
+   *   <li>CHAR --> ArrowType.Utf8</li>
+   *   <li>NCHAR --> ArrowType.Utf8</li>
+   *   <li>VARCHAR --> ArrowType.Utf8</li>
+   *   <li>NVARCHAR --> ArrowType.Utf8</li>
+   *   <li>LONGVARCHAR --> ArrowType.Utf8</li>
+   *   <li>LONGNVARCHAR --> ArrowType.Utf8</li>
+   *   <li>NUMERIC --> ArrowType.Decimal(precision, scale)</li>
+   *   <li>DECIMAL --> ArrowType.Decimal(precision, scale)</li>
+   *   <li>BIT --> ArrowType.Bool</li>
+   *   <li>TINYINT --> ArrowType.Int(8, signed)</li>
+   *   <li>SMALLINT --> ArrowType.Int(16, signed)</li>
+   *   <li>INTEGER --> ArrowType.Int(32, signed)</li>
+   *   <li>BIGINT --> ArrowType.Int(64, signed)</li>
+   *   <li>REAL --> ArrowType.FloatingPoint(FloatingPointPrecision.SINGLE)</li>
+   *   <li>FLOAT --> ArrowType.FloatingPoint(FloatingPointPrecision.SINGLE)</li>
+   *   <li>DOUBLE --> ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)</li>
+   *   <li>BINARY --> ArrowType.Binary</li>
+   *   <li>VARBINARY --> ArrowType.Binary</li>
+   *   <li>LONGVARBINARY --> ArrowType.Binary</li>
+   *   <li>DATE --> ArrowType.Date(DateUnit.MILLISECOND)</li>
+   *   <li>TIME --> ArrowType.Time(TimeUnit.MILLISECOND, 32)</li>
+   *   <li>TIMESTAMP --> ArrowType.Timestamp(TimeUnit.MILLISECOND, calendar timezone)</li>
+   *   <li>CLOB --> ArrowType.Utf8</li>
+   *   <li>BLOB --> ArrowType.Binary</li>
+   * </ul>
+   *
+   * @param fieldInfo The field information to construct the <code>ArrowType</code> from.
+   * @param calendar The calendar to use when constructing the <code>ArrowType.Timestamp</code>
+   *                 for {@link java.sql.Types#TIMESTAMP} types.
+   * @return The corresponding <code>ArrowType</code>.
+   * @throws NullPointerException if either <code>fieldInfo</code> or <code>calendar</code> are <code>null</code>.
+   */
+  public static ArrowType getArrowTypeForJdbcField(JdbcFieldInfo fieldInfo, Calendar calendar) {
```

Comment:
you could potentially move this to a method on fieldInfo

Reply:
I wasn't trying to stray too far from what the original author was doing, since his work did pass code review ...

The diff is a little confusing, but the original logic is on the old logic is on the old (red) line 249.